### PR TITLE
Add developers notes

### PR DESCRIPTION
**Pull request by JayGreeeen**
### :pencil: Description
I've added a DEVELOPERS.md file based off of the notes I have made while working on BigQuery. I have also added a logo for the project which is an adaptation of the existing Circus Train logo with the addition of the Google BigQuery logo. 

### :link: Related Issues

### CODE REVIEW

File: DEVELOPERS.md [hunk 1]
```diff
@@ -0,0 +1,175 @@
+![Circus Train BigQuery.](circus-train-bigquery.png "Moving data from Google BigQuery to Hive.")
+
+## Overview 
+Circus Train BigQuery is an extension of Circus Train which allows tables in Google BigQuery to be replicated.
+
+This file contains a collection of notes put together by developers who have worked on BigQuery to provide some explanation to the code. This file is not completely exhaustive of all the inner workings of the project, so do feel free to add more information or detail.
```
Comment (massdosage) [inline DEVELOPERS.md:6]:
```suggestion
This document contains a collection of notes put together by developers who have worked on the project to provide some explanations of the code and how it works. This is not completely exhaustive of all the inner workings of the project, so do feel free to add more information or detail. 
```

File: DEVELOPERS.md [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+
+## Overview 
+Circus Train BigQuery is an extension of Circus Train which allows tables in Google BigQuery to be replicated.
+
+This file contains a collection of notes put together by developers who have worked on BigQuery to provide some explanation to the code. This file is not completely exhaustive of all the inner workings of the project, so do feel free to add more information or detail. 
+
+
+## README.md
+
+First and foremost, its worth having a read through the Circus Train [README.md](https://github.com/HotelsDotCom/circus-train) file and the Circus Train BigQuery [README.md](https://github.com/HotelsDotCom/circus-train-bigquery/blob/master/README.md). These are pretty extensive guides containing a lot of info on the projects, including how to run each of them and the different configurations which can be used.
```
Comment (massdosage) [inline DEVELOPERS.md:11]: Link directly to the CT README.md like you do for the other one.

File: DEVELOPERS.md [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+Circus Train BigQuery is an extension of Circus Train which allows tables in Google BigQuery to be replicated.
+
+This file contains a collection of notes put together by developers who have worked on BigQuery to provide some explanation to the code. This file is not completely exhaustive of all the inner workings of the project, so do feel free to add more information or detail. 
+
+
+## README.md
+
+First and foremost, its worth having a read through the Circus Train [README.md](https://github.com/HotelsDotCom/circus-train) file and the Circus Train BigQuery [README.md](https://github.com/HotelsDotCom/circus-train-bigquery/blob/master/README.md). These are pretty extensive guides containing a lot of info on the projects, including how to run each of them and the different configurations which can be used. 
+
+It may also be useful to read through the DEVELOPERS.md file for Circus Train.
```
Comment (massdosage) [inline DEVELOPERS.md:13]: Make this a link too (you can figure out what the location should be, but we should merge that other PR first).

File: DEVELOPERS.md [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+
+## README.md
+
+First and foremost, its worth having a read through the Circus Train [README.md](https://github.com/HotelsDotCom/circus-train) file and the Circus Train BigQuery [README.md](https://github.com/HotelsDotCom/circus-train-bigquery/blob/master/README.md). These are pretty extensive guides containing a lot of info on the projects, including how to run each of them and the different configurations which can be used. 
+
+It may also be useful to read through the DEVELOPERS.md file for Circus Train. 
+
+
+## Basic description
+Circus Train BigQuery handles applying any partition filters to the data in the Google BigQuery table and extracts it as Avro into Google Cloud Storage, (GCS). Circus Train listeners are used to convert the Google BigQuery metadata into a Hive table object. The data is then replicated from source to target using the metadata from this mocked Hive table.
```
Comment (massdosage) [inline DEVELOPERS.md:17]: I think we first need a very high level description, then flesh it out with some of the additional info you have here. Something like "At a high level CTBQ uses X integration points provided by CT (mention them, the copiers or whatever) to replicate data out of BigQuery and into Hive. It first runs a query to extract the data out of BQ into a temporary BQ table (applying any necessary partition filters) and then uses the BQ X API (find out which one the extractor uses) to extract that data out of this temporary table onto GCS as Avro files. A Hive table _object_ (i.e. not a real Hive table) is then put "on top" of this data and Circus train is instructed to replicate the data from this source to the target as normal."

File: DEVELOPERS.md [hunk 5]
```diff
@@ -0,0 +1,175 @@
...
+
+If the table is **partitioned**:
```
Comment (massdosage) [inline DEVELOPERS.md:19]:
```suggestion
### Partitioned Tables
```

File: DEVELOPERS.md [hunk 6]
```diff
@@ -0,0 +1,175 @@
...
+
+* A temporary table is created by applying a partition filter to the source table. 
+* The data from this temporary table is copied over to GCS.
```
Comment (massdosage) [inline DEVELOPERS.md:22]:
```suggestion
* The data from this temporary table extracted to GCS. 
```

File: DEVELOPERS.md [hunk 7]
```diff
@@ -0,0 +1,175 @@
...
+Circus Train BigQuery handles applying any partition filters to the data in the Google BigQuery table and extracts it as Avro into Google Cloud Storage, (GCS). Circus Train listeners are used to convert the Google BigQuery metadata into a Hive table object. The data is then replicated from source to target using the metadata from this mocked Hive table.
+
+If the table is **partitioned**:
+
+* A temporary table is created by applying a partition filter to the source table. 
+* The data from this temporary table is copied over to GCS. 
+* Circus Train then uses this GCS location to perform the replication to Hive.
+
+
+If the table is **unpartitioned**:
```
Comment (massdosage) [inline DEVELOPERS.md:26]:
```suggestion
### Unpartitioned Tables
```

File: DEVELOPERS.md [hunk 8]
```diff
@@ -0,0 +1,175 @@
...
+
+* The data from the source table is copied over to GCS.
```
Comment (massdosage) [inline DEVELOPERS.md:28]:
```suggestion
* The data from the entire source table is extracted to GCS.
```

File: DEVELOPERS.md [hunk 9]
```diff
@@ -0,0 +1,175 @@
...
+`<name>` - Lowercase names indicate methods
+
+
+### Classes
+
+**Locomotive** (CT)
+
+* Creates a new `Replication` (CT) object using the `ReplicationFactory` (CT) and calls `replicate` on it.
+
+**ReplicationFactor** (CT)
```
Comment (massdosage) [inline DEVELOPERS.md:53]:
```suggestion
**ReplicationFactory* (CT)
```

File: DEVELOPERS.md [hunk 10]
```diff
@@ -0,0 +1,175 @@
...
+**Replication** (CT)
+
+* Either `PartitionedTableReplication` (CT) or `UnpartitionedTableReplication` (CT).
+* Uses the `BigQueryCopierFactory` to generate the `BigQueryCopier`, then calls `copy` to copy over the table data. 
+* After that will update the metadata of the table. 
+
+
+**BigQueryMetastoreClientFactory**
+
+* Checks the given uri is acceptable for BigQuery.
```
Comment (massdosage) [inline DEVELOPERS.md:67]:
```suggestion
* Checks the given URI is acceptable for BigQuery.
```

File: DEVELOPERS.md [hunk 11]
```diff
@@ -0,0 +1,175 @@
...
+* Returns a new instance of the `BigQueryMetastoreClient`.
+
+
+**BigQueryMetastoreClient**
+
+* Registers a container with the `ExtractionService`.
+* Container has an `UpdateTableSchemaAction` to be executed by the `ExtractionService`.
+* Returns a hive table - based on the BigQuery table. Creates the cached table object.
```
Comment (massdosage) [inline DEVELOPERS.md:75]:
```suggestion
* Returns a Hive table - based on the BigQuery table. Creates the cached table object.
```

File: DEVELOPERS.md [hunk 12]
```diff
@@ -0,0 +1,175 @@
...
+
+
+**BigQueryMetastore**
+
+* Executes the partition filter query and adds the results to a temporary table in BigQuery.
+
+
+**TableServiceFactory**
+
+* Will create either a `PartitionedTableService` or an `UnpartitionedTableService`, based on whether the hive table being replicated is partitioned or not.
```
Comment (massdosage) [inline DEVELOPERS.md:85]:
```suggestion
* Will create either a `PartitionedTableService` or an `UnpartitionedTableService`, based on whether the Hive table being replicated is partitioned or not.
```
(check rest of document for upper case H in Hive)

File: DEVELOPERS.md [hunk 13]
```diff
@@ -0,0 +1,175 @@
...
+**Composite Copier Factory** (CT)
+
+* Creates the `BigQueryCopierFactory` instance.
+* Adds it to a `CompositeCopier` object.
+* Later in `PartitionedTableReplication` or `UnpartitionedTableReplication`, copy will be called on the `CompositeCopier` - which will call copy on the `BQCopier`.
+
+
+**BigQueryCopierFactory**
+
+* Creates the `BigQueryCopier`
```
Comment (massdosage) [inline DEVELOPERS.md:131]:
```suggestion
* Creates the `BigQueryCopier`.
```

File: DEVELOPERS.md (hunk 3 above)
Comment (JayGreeeen) [reply DEVELOPERS.md:13]: Yeah I will do 👍

File: DEVELOPERS.md [hunk 14]
```diff
@@ -1,20 +1,22 @@
...
 
 This document contains a collection of notes put together by developers who have worked on the project to provide some explanations of the code and how it works. This is not completely exhaustive of all the inner workings of the project, so do feel free to add more information or detail. 
 
 
 ## README.md
 
-First and foremost, its worth having a read through the Circus Train [README.md](https://github.com/HotelsDotCom/circus-train) file and the Circus Train BigQuery [README.md](https://github.com/HotelsDotCom/circus-train-bigquery/blob/master/README.md). These are pretty extensive guides containing a lot of info on the projects, including how to run each of them and the different configurations which can be used. 
+First and foremost, its worth having a read through the Circus Train [README.md](https://github.com/HotelsDotCom/circus-train/blob/master/README.md) file and the Circus Train BigQuery [README.md](https://github.com/HotelsDotCom/circus-train-bigquery/blob/master/README.md). These are pretty extensive guides containing a lot of info on the projects, including how to run each of them and the different configurations which can be used.
```
Comment (barnharts4) [inline DEVELOPERS.md:11]:
```suggestion
First and foremost, it's worth having a read through the Circus Train [README.md](https://github.com/HotelsDotCom/circus-train/blob/master/README.md) file and the Circus Train BigQuery [README.md](https://github.com/HotelsDotCom/circus-train-bigquery/blob/master/README.md). These are pretty extensive guides containing a lot of info on the projects, including how to run each of them and the different configurations which can be used. 
```

File: DEVELOPERS.md [hunk 15]
```diff
@@ -1,20 +1,22 @@
...
+It may also be useful to read through the [DEVELOPERS.md](https://github.com/HotelsDotCom/circus-train/blob/master/DEVELOPERS.md) file for Circus Train. 
 
 
 ## Basic description
-Circus Train BigQuery handles applying any partition filters to the data in the Google BigQuery table and extracts it as Avro into Google Cloud Storage, (GCS). Circus Train listeners are used to convert the Google BigQuery metadata into a Hive table object. The data is then replicated from source to target using the metadata from this mocked Hive table.
+
+At a high level, CTBQ uses the Replication and Copier integration points provided by CT to replicate data out of BigQuery and into Hive. It first runs a query to extract the data out of BQ into a temporary BQ table (applying any necessary partition filters) and then uses Google's BigQuery and Storage APIs to create a Job which extracts that data out of this temporary table onto Google Cloud Storage, GCS, as Avro files. A Hive table object (i.e. not a real Hive table) is then put "on top" of this data and Circus train is instructed to replicate the data from this GCS source to the target as normal.
```
Comment (barnharts4) [inline DEVELOPERS.md:18]:
```suggestion
At a high level, CTBQ uses the Replication and Copier integration points provided by CT to replicate data out of BigQuery and into Hive. It first runs a query to extract the data out of BQ into a temporary BQ table (applying any necessary partition filters) and then uses Google's BigQuery and Storage APIs to create a Job which extracts that data out of this temporary table onto Google Cloud Storage, GCS, as Avro files. A Hive table object (i.e. not a real Hive table) is then put "on top" of this data and Circus Train is instructed to replicate the data from this GCS source to the target as normal.
```

Comment (barnharts4) [reply DEVELOPERS.md:18]: Hard to see the difference - the last instance of `Circus train` was changed to `Circus Train`.

**Comment by JayGreeeen** [comment]
@massdosage @barnharts4 thanks for all your help with both of these PRs, very much appreciated!

**Pull request merged**